keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/ForwardSampler.java

```diff
@@ -0,0 +1,114 @@
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    public NetworkSamples sample(BayesianNetwork network,
+                                        List<? extends Variable> fromVariables,
+                                        int sampleCount) {
+        return sample(network, fromVariables, sampleCount, KeanuRandom.getDefaultRandom());
+    }
+
+    public NetworkSamples sample(BayesianNetwork network,
+                                        List<? extends Variable> fromVariables,
+                                        int sampleCount,
+                                        KeanuRandom random) {
+
+        Preconditions.checkArgument(network.getObservedVertices().size() == 0, "Cannot forward sample from a network with observations");
+
+        List<Vertex> fromVertices = fromVariables.stream().map(v -> (Vertex) v).collect(Collectors.toList());
+
+        Set<Vertex> probabilisticSubset = fromVertices.stream().filter(v -> v instanceof Probabilistic).collect(Collectors.toSet());
+        Set<Vertex> nonProbabilisticSubset = fromVertices.stream().filter( v -> !(v instanceof Probabilistic)).collect(Collectors.toSet());
+
+        upstreamOfProbabilisticDoesNotContainProbabilistic(probabilisticSubset);
+
+        ComputableModel graph = new KeanuComputableModel(new HashSet<>(network.getAllVertices()));
+        KeanuProposalApplicationStrategy proposalApplicationStrategy = new KeanuProposalApplicationStrategy(probabilisticSubset);
+
+        return sample(graph, probabilisticSubset, nonProbabilisticSubset, proposalApplicationStrategy, sampleCount, random);
+    }
+
+    private NetworkSamples sample(ComputableModel graph,
+                                        Set<? extends Variable> probabilisticFromVariables,
+                                        Set<? extends Variable> nonProbabilisticFromVariables,
+                                        ProposalApplicationStrategy proposalApplicationStrategy,
+                                        int sampleCount,
+                                        KeanuRandom random) {
+
+        Set<? extends Variable> allSampleFromVariables = Sets.union(probabilisticFromVariables, nonProbabilisticFromVariables);
+        Map<VariableReference, List> samplesByVariable = new HashMap<>();
+
+        for (int sampleNum = 0; sampleNum < sampleCount; sampleNum++) {
+            Proposal probabilisticVariableProposal = PRIOR_PROPOSAL.getProposal(probabilisticFromVariables, random);
+            proposalApplicationStrategy.apply(probabilisticVariableProposal);
+            graph.compute(Collections.EMPTY_MAP, nonProbabilisticFromVariables.stream().map(Variable::getReference).collect(Collectors.toList()));
+            takeSamples(samplesByVariable, allSampleFromVariables);
+        }
+
+        ArrayList<Double> logProb = new ArrayList<>(Collections.nCopies(sampleCount, LOG_PROB_OF_PRIOR));
```

Comment:
Is this correct?  Surely samples from the Prior still have a logProb for the likelihood?

Reply:
The `likelihood` is the probability of the observations. Do we have a use case where we need the logProb? I think this sampler violates our previous assumption that all samples from the posterior need a logProb. We only use this logProb of the samples for determining if the samples have converged on the posterior but in this case we know we're sampling directly from the posterior. We would need the logProb for VI though so maybe make it optional?

Reply:
Yeah - I believe this was initially added for KL divergence calculations and the like.  As you say given this sampling algorithm doesn't need to calculate the logProb as part of the operation it would be best to make this optional (and probably follow through with the other sampling algorithms too).  Perhaps something to raise a ticket for?